Support Subaru FOCAS

ejeschke opened the pull request:
Adds support for the Subaru FOCAS spectrograph.

profxj committed:
viewfits

profxj committed:
mo

profxj committed:
wavelengths

ejeschke committed:
Correct lookup of Subaru Telescope location

ejeschke committed:
Changes suggested by @Prof_X and @Debora

- change to TellPCA_3000_10500_R120000.fits

profxj committed:
quadratic fix

profxj committed:
mo

profxj committed:
WIP

ejeschke committed:
Fixes and updates for Subaru FOCAS

- corrected noise levels for all amps and detectors
- corrected pixel (plate) scale
- added saturation and nonlinear parameters
- change exposure range for arc frames
- added get_rawimage() method to remove overscan regions

ejeschke committed:
Correct dark current values

ejeschke committed:
Update exposure ranges for science and standard frames

- both use OBJECT for the DATA-TYP keyword
- TODO: come up with a better way to distinguish between them

ejeschke commented:
NOTE: rebased against `develop` using the current branch of "subaru_focas" at Pypeit.

profxj commented on pypeit/core/wavecal/wvutils.py:
```diff
@@ -775,10 +775,14 @@ def xcorr_shift_stretch(inspec1, inspec2, cc_thresh=-1.0, percent_ceil=50.0, use
             bounds = [lag_range, stretch_mnmx, (-1.0e-6, 1.0e-6)]
             x0_guess = np.array([shift_cc, 1.0, 0.0])
-        if stretch_func == 'linear':
+        elif stretch_func == 'linear':
```
good catch!

profxj commented on pypeit/spectrographs/subaru_focas.py:
```diff
@@ -0,0 +1,688 @@
+
+
+    # TODO -- Convert this into get_comb_group()
+    def parse_dither_pattern(self, file_list, ext=None):
```
do we actually use this?  if not, please remove.

ejeschke replied:
@Tang-SL , did you add this? Pretty sure our team did not, as we were not addressing dithers.

Tang-SL replied:
@ejeschke Nope, that wasn't me.

ejeschke replied:
Great, thanks for chiming in.  I'll remove it then.

ejeschke replied:
Removed

profxj commented on pypeit/spectrographs/subaru_focas.py:
```diff
@@ -0,0 +1,688 @@
+                'SLT-WID']  #, 'SLTCPIX1', 'SLTCPIX2', 'SLTC-RA', 'SLTC-DEC']
+
+
+# Definitions for the over scan regions in the DS9 image coordinate.
```
are these used somewhere?  if so, wrap them in a `def`
If not, please remove

ejeschke replied:
Wrapped!

profxj commented on pypeit/telescopes.py:
```diff
@@ -44,6 +44,17 @@ def __init__(self):
+                                               elevation=loc.height.to(units.m).value,
+                                               fratio=12.2,  # CASSEGRAIN
+                                               diameter=8.2,
+                                               eff_aperture=72.3674) # TODO -- FIX THIS??
```
is the `eff_aperture` ok?  if so, remove the TODO
  if not, fix it  :)

ejeschke replied:
I need to find out whether this value is reasonably correct. Asking around...

profxj reviewed (changes requested):
just a few things from me

thanks!

ejeschke committed:
Clean up some dead code and add arc line files

profxj commented:
please pull in `develop`

ejeschke commented:
> please pull in `develop`

Rebased against `develop` again

debora-pe commented on pypeit/spectrographs/subaru_focas.py:
```diff
@@ -0,0 +1,549 @@
+
+        # Median overscan
+        #   IF YOU CHANGE THIS, YOU WILL NEED TO DEAL WITH THE OVERSCAN GOING ALONG ROWS
+        for key in par['calibrations'].keys():
```
can this be simply `par.reset_all_processimages_par(overscan_method='median')`?
Take a look at examples in, e.g., keck_hires and lbt_mods.

profxj replied:
good call; implemented

debora-pe commented on pypeit/spectrographs/subaru_focas.py:
```diff
@@ -0,0 +1,549 @@
+        self.meta['decker'] = dict(ext=0, card='SLIT')
+        # Extras for config and frametyping
+        self.meta['dispname'] = dict(ext=0, card='DISPERSR', required_ftypes=['science', 'standard'])
+        # TODO - FIX THIS!!
```
Is this true? Does it needs to be fixed?

profxj replied:
TODO removed

debora-pe commented on pypeit/spectrographs/subaru_focas.py:
```diff
@@ -0,0 +1,549 @@
+        return (detector, rawdata, hdu_l, exptime, rawdatasec_img, oscansec_img)
+
+
+    @property
```
remove this function

debora-pe commented on pypeit/spectrographs/subaru_focas.py:
```diff
@@ -0,0 +1,549 @@
+        return [(1,2)]
+
+    @property
+    def default_mosaic(self):
```
remove this function

debora-pe commented on pypeit/telescopes.py:
```diff
@@ -44,6 +44,17 @@ def __init__(self):
+                                               elevation=loc.height.to(units.m).value,
+                                               fratio=12.2,  # CASSEGRAIN
+                                               diameter=8.2,
+                                               eff_aperture=72.3674) # TODO -- FIX THIS??
```
Just a reminder for this `eff_aperture`

debora-pe reviewed (changes requested):
A few changes requested.
Some documentation on FOCAS is also appreciated. See an example here: `PypeIt/doc/spectrographs/lris.rst` or other files in that folder.
Also, a reminder that we'll need a test data in the Dev-Suite.

ejeschke committed:
Restore some conditionals for various grisms

ejeschke commented:
Updated and rebased against `develop`.  I think I addressed most of the comments here for the code, except documentation.  I think there might be some more reid_arxiv arc files to add too.

ejeschke committed:
Add wavelength solution for SCFCGRHD85 grism

ejeschke committed:
Add XX1 as bit type for wavecal

- seems to be necessary, even though we renamed all the XX1 types
  to UNKNWN in pypeit/data/arc_lines/lists/ThAr_lines.dat
  Not sure where this is being picked up
